### DISCUSSION TITLE
Docker error

### DISCUSSION THREAD

**Discussion by Lure5134** [QuestPDF/QuestPDF#873] [Q&A]
I get the following error in the newest version of QuestPDF when running it inside docker on a linux machine.

> Unhandled exception. System.TypeInitializationException: The type initializer for 'QuestPDF.Settings' threw an exception.
 ---> System.Exception: The QuestPDF library has encountered an issue while loading one of its dependencies.
 ---> System.DllNotFoundException: Unable to load shared library 'QuestPdfSkia' or one of its dependencies. In order to help diagnose loading problems, consider using a tool like strace. If you're using glibc, consider setting the LD_DEBUG environment variable: 
/usr/share/dotnet/shared/Microsoft.NETCore.App/8.0.4/QuestPdfSkia.so: cannot open shared object file: No such file or directory
/app/QuestPdfSkia.so: cannot open shared object file: No such file or directory
/usr/share/dotnet/shared/Microsoft.NETCore.App/8.0.4/libQuestPdfSkia.so: cannot open shared object file: No such file or directory
/app/libQuestPdfSkia.so: cannot open shared object file: No such file or directory
/usr/share/dotnet/shared/Microsoft.NETCore.App/8.0.4/QuestPdfSkia: cannot open shared object file: No such file or directory
/app/QuestPdfSkia: cannot open shared object file: No such file or directory
/usr/share/dotnet/shared/Microsoft.NETCore.App/8.0.4/libQuestPdfSkia: cannot open shared object file: No such file or directory
/app/libQuestPdfSkia: cannot open shared object file: No such file or directory
   at QuestPDF.Skia.SkNativeDependencyCompatibilityChecker.API.check_compatibility_by_calculating_sum(Int32 a, Int32 b)
   at QuestPDF.Skia.SkNativeDependencyCompatibilityChecker.CheckIfExceptionIsThrownWhenLoadingNativeDependencies() in /src/ref/QuestPDF/Source/QuestPDF/Skia/SkNativeDependencyCompatibilityChecker.cs:line 69
   --- End of inner exception stack trace ---
   at QuestPDF.Skia.SkNativeDependencyCompatibilityChecker.Test() in /src/ref/QuestPDF/Source/QuestPDF/Skia/SkNativeDependencyCompatibilityChecker.cs:line 56
   at QuestPDF.Helpers.NativeDependencyCompatibilityChecker.Test() in /src/ref/QuestPDF/Source/QuestPDF/Helpers/NativeDependencyCompatibilityChecker.cs:line 19
   at QuestPDF.Settings..cctor() in /src/ref/QuestPDF/Source/QuestPDF/Settings.cs:line 55
   --- End of inner exception stack trace ---
   at QuestPDF.Settings.set_License(Nullable`1 value) in /src/ref/QuestPDF/Source/QuestPDF/Settings.cs:line 13

When I inspect the docker image these files exists containing the word "Skia"

> app/Runtimes/linux-arm64/native/libQuestPdfSkia.so
app/Runtimes/linux-musl-x64/native/libQuestPdfSkia.so
app/Runtimes/linux-x64/native/libQuestPdfSkia.so
app/Runtimes/osx-arm64/native/libQuestPdfSkia.dylib
app/Runtimes/osx-x64/native/libQuestPdfSkia.dylib
app/Runtimes/win-x64/native/QuestPdfSkia.dll

My Dockerfile looks like this

> FROM mcr.microsoft.com/dotnet/aspnet:8.0 AS base
WORKDIR /app
EXPOSE 80
EXPOSE 443
FROM mcr.microsoft.com/dotnet/sdk:8.0 AS build
WORKDIR /src
COPY TestProject/*.csproj TestProject/
RUN dotnet restore -r linux-x64 "TestProject/TestProject.csproj"
COPY . .
WORKDIR "/src/TestProject"
RUN dotnet build "TestProject.csproj" -r linux-x64 -c Release -o /app/build
FROM build AS publish
RUN dotnet publish "TestProject.csproj" -r linux-x64 -c Release -o /app/publish
FROM base AS final
WORKDIR /app
COPY --from=publish /app/publish .
ENTRYPOINT ["dotnet", "TestProject.dll"]

**Comment by Lure5134** [comment]
@MarcinZiabek I have tested it again with the newest version and still get this error.

> Unhandled exception. System.TypeInitializationException: The type initializer for 'QuestPDF.Settings' threw an exception.
 ---> System.Exception: The QuestPDF library has encountered an issue while loading one of its dependencies.
 ---> System.DllNotFoundException: Unable to load shared library 'QuestPdfSkia' or one of its dependencies. In order to help diagnose loading problems, consider using a tool like strace. If you're using glibc, consider setting the LD_DEBUG environment variable: 
/usr/share/dotnet/shared/Microsoft.NETCore.App/8.0.4/QuestPdfSkia.so: cannot open shared object file: No such file or directory
/app/QuestPdfSkia.so: cannot open shared object file: No such file or directory
/usr/share/dotnet/shared/Microsoft.NETCore.App/8.0.4/libQuestPdfSkia.so: cannot open shared object file: No such file or directory
/app/libQuestPdfSkia.so: cannot open shared object file: No such file or directory
/usr/share/dotnet/shared/Microsoft.NETCore.App/8.0.4/QuestPdfSkia: cannot open shared object file: No such file or directory
/app/QuestPdfSkia: cannot open shared object file: No such file or directory
/usr/share/dotnet/shared/Microsoft.NETCore.App/8.0.4/libQuestPdfSkia: cannot open shared object file: No such file or directory
/app/libQuestPdfSkia: cannot open shared object file: No such file or directory
   at QuestPDF.Skia.SkNativeDependencyCompatibilityChecker.API.check_compatibility_by_calculating_sum(Int32 a, Int32 b)
   at QuestPDF.Skia.SkNativeDependencyCompatibilityChecker.CheckIfExceptionIsThrownWhenLoadingNativeDependencies() in /src/ref/QuestPDF/Source/QuestPDF/Skia/SkNativeDependencyCompatibilityChecker.cs:line 72
   --- End of inner exception stack trace ---
   at QuestPDF.Skia.SkNativeDependencyCompatibilityChecker.Test() in /src/ref/QuestPDF/Source/QuestPDF/Skia/SkNativeDependencyCompatibilityChecker.cs:line 59
   at QuestPDF.Helpers.NativeDependencyCompatibilityChecker.Test() in /src/ref/QuestPDF/Source/QuestPDF/Helpers/NativeDependencyCompatibilityChecker.cs:line 19
   at QuestPDF.Settings..cctor() in /src/ref/QuestPDF/Source/QuestPDF/Settings.cs:line 55
   --- End of inner exception stack trace ---

Do you know what could be the problem?
I use QuestPDF as a submodule and not as a nuget package. Could that be the problem?

**Reply by MarcinZiabek** [reply]
I also think that using the library as a submodule is a likely root cause. QuestPDF uses `nuspec` files to copy necessary native artifacts to your bin folder. Nuspec files are usually executed during the `nuget restore` process, of course for nuget packages.

**Comment by DawidRobertNowak** [comment]
Hello,

We are closing this thread due to inactivity. We apologize for the delayed response. Thank you for taking the time to reach out and contribute to the project — your feedback helps us improve QuestPDF.

Since this thread was opened, we have done significant work to improve QuestPDF's compatibility across various Linux environments, including support for older glibc versions and the `linux-x64`, `linux-arm64`, and `linux-musl-x64` runtimes. We recommend updating to a recent version, which should resolve native dependency loading issues like this one in Docker.

Should you encounter any further problems, please don't hesitate to open a new issue.
Thank you!
